### DISCUSSION TITLE
yggdrasilctl addpeer understanding

### DISCUSSION THREAD

**Discussion by d47081** [yggdrasil-network/yggdrasil-go#1054] [Q&A] [answered]
how can I properly register new peer by `yggdrasilctl addpeer` command

as the argument, I've provided peer ipv6 address, clearnet IP, schema including and everything returns an error:

```
Admin socket returned an error: unknown call scheme:
```

could someone help with command example or link to documentation?
thanks!

**Comment by d47081** [comment]
maybe this option means 'public' peers  only?

I just want to connect remote (non-public) node to another one where I can't find it in `getPeers `output

**Reply by xialvjun** [reply]
Both A and B are behind NAT network. If A want to connect to B, A need to connect to a node(let's say C) with public ip directly or indirectly(A -> C or A -> X -> C), and B is the same(B -> C or B -> Y -> C). 

Choose a public peer in https://publicpeers.neilalexander.dev/ and add it into your `yggdrasil.conf -> Peers`. 

You can believe these public peers have connected to each other. So A and B can choose different public peer.

**Reply by d47081** [reply]
thanks for reply,

sure I have connected nodes using public peer (one of them maintaining also)

but when launching the B node, it have no A in `yggdrasilctl getPeers` list, so then trying to execute `yggdrasilctl addPeer ip/public key/etc` of A and receiving  error
```
Admin socket returned an error: unknown call scheme:
```

so asking about this command maybe just to connect public peers only, and I can't forcefully discover another one by it ip or pubkey providing.

I can connect A <-> B using 200/7 ipv6 but I can't see A/B peers in `yggdrasilctl getPeers`

**Comment by d47081** [comment]
To explain my subject in another words, I can't find some nodes in [YGGstate explorer](https://github.com/YGGverse/YGGstate)

and want manually register new peer to track it stats there 
(or implement peer auto-registration on search request when not found yet in the DB, even online)

**Reply by xialvjun** [reply]
I don't know if I'm right. 

I think `yggdrasilctl getPeers` is things about `real-network`. Our 200/7 ipv6 is `yggdrasil-network`. They are different things in different layer.

These yggdrasil ipv6 addresses are static ip, so we need static data to show the network topology. We can not get a self-discovery network topology. 

Of course if we run `yggdrasilctl getself` in every node, and send the output to a server, that's a self-discovery network topology.

**Reply by d47081** [reply]
usually, in same DHT networks it's possible to provide real IP address with open port, where peer B instance running, and peer A adds it to the known peers list, and successfully connects. 

but with `yggdrasilctl addpeer peerIPv4` it returns message above, maybe forgot to provide the port or so, will try again this evening...

anyway, thanks for thoughts

**Comment by d47081** [comment, marked as answer]
Seems, that command operate with [public peers](https://publicpeers.neilalexander.dev/) only.
and after restarting the service, yggdrasil will load default public peers, provided in yggdrasil.conf.

my command have returned error, because I've missed to provide connection protocol, and uri/interface prefix:

`yggdrasilctl list`

```
addpeer uri=..., interface=... Add a peer to the peer list
```

yggdrasil.conf

```
  # List of connection strings for outbound peer connections in URI format,
  # e.g. tls://a.b.c.d:e or socks://a.b.c.d:e/f.g.h.i:j. These connections
  # will obey the operating system routing table, therefore you should
  # use this section when you may connect via different interfaces.
  Peers: []

  # List of connection strings for outbound peer connections in URI format,
  # arranged by source interface, e.g. { "eth0": [ "tls://a.b.c.d:e" ] }.
  # Note that SOCKS peerings will NOT be affected by this option and should
  # go in the "Peers" section instead.
  InterfacePeers: {}
```

And forgot to say, have tried to connect A and B nodes by it direct IPs, that for test, was connected to different public peers.

Just wondering, why even node website accessible over NAT, I can't discover this node connection by it unique ipv6 in the swarm.

**Comment by d47081** [comment]
Well, seems that public peer have no more data than connected peers of. 
I've just added an invite message to connect public peer of YGGstate tracker  on search results page.

It births the idea make stats data federated by API.

My subject requires alternative solutions. Replied on all questions above, so closing the subject.
